pgml-cms/blog/semantic-search-in-postgres-in-15-minutes.md

```diff
@@ -0,0 +1,467 @@
+ 0.1114425936213167
+
+cosine_distance   
+--------------------
+ 0.7328613577628744
+```
+
+!!!
+
+!!!
+
+You'll notice that the distance between "What is the pgml.transform function?" and "The pgml.transform function is a PostgreSQL function for calling LLMs in the database." is much smaller than the cosine distance between "What is the pgml.transform function?" and "I think tomatoes are incredible on burgers".
+
+## Making it fast!
+
+It is inefficient to compute embeddings for all the documents every time we search the dataset. Instead, we should embed our documents once and search against precomputed embeddings. 
+
+_pgvector_ provides us with the `vector` data type for storing embeddings in regular PostgreSQL tables:
+
+!!! generic
+
+!!! code_block
+
+```postgresql
+CREATE TABLE text_and_embeddings (
+    id SERIAL PRIMARY KEY, 
+    text text, 
+    embedding vector (1024)
+);
+
+INSERT INTO text_and_embeddings (text, embedding)
+VALUES 
+  (
+    'The pgml.transform function is a PostgreSQL function for calling LLMs in the database.',
+    pgml.embed(
+      'mixedbread-ai/mxbai-embed-large-v1',
+      'The pgml.transform function is a PostgreSQL function for calling LLMs in the database.'
+    )
+  ),
+
+  (
+    'I think tomatoes are incredible on burgers.',
+    pgml.embed(
+      'mixedbread-ai/mxbai-embed-large-v1',
+      'I think tomatoes are incredible on burgers.'
+    )
+  );
+```
+
+!!!
+
+!!!
+
+Once our table has some data, we can search it using the following query: 
+
+!!! generic
+
+!!! code_block time="19.864 ms"
+
+```postgresql
+WITH query_embedding AS (
+    SELECT
+        pgml.embed(
+          'mixedbread-ai/mxbai-embed-large-v1',
+          'What is the pgml.transform function?',
+          '{"prompt": "Represent this sentence for searching relevant passages: "}'
+        )::vector embedding
+)
+SELECT
+    text,
+    (
+        SELECT
+          embedding
+        FROM query_embedding
+    ) <=> text_and_embeddings.embedding cosine_distance
+FROM
+  text_and_embeddings
+ORDER BY cosine_distance
+LIMIT 1;
+```
+
+!!!
+
+!!! results
+
+```
+                                          text                                          |   cosine_distance   
+----------------------------------------------------------------------------------------+---------------------
+ The pgml.transform function is a PostgreSQL function for calling LLMs in the database. | 0.13467974993681486
+```
+
+!!!
+
+!!!
+
+This query is fast for now, but as we add more data to the table, it will slow down because we have not indexed the embedding column. 
+
+Let's demonstrate this by inserting 100,000 additional embeddings:
+
+!!! generic
+
+!!! code_block
+
+```postgresql
+INSERT INTO text_and_embeddings (text, embedding) 
+SELECT
+  md5(random()::text),
+  pgml.embed(
+    'mixedbread-ai/mxbai-embed-large-v1',
+    md5(random()::text)
+  ) 
+FROM generate_series(1, 100000);
+```
+
+!!!
+
+!!!
+
+Now trying our search engine again:
```

Comment:
This would be better illustrated with a graph at 10, 100, 1000, 10000, 100k, 1M. Show that Big O.